third_party/nvfuser/csrc/scheduler/matmul.cpp

```diff
@@ -320,4 +329,305 @@ void scheduleMatmul(
+      MmaBuilder(params.mma_op, params.tile_sizes).layout(params.layout);
+  const auto& gemm_tile = params.tile_sizes;
+
+  // Including current tensor naming convention for reference,
+  //  this is very temporary and will change over time and
+  //  in fact the whole body of this function will
+  //  eventually be a set of utility functions for different
+  //  sections of matmul(fusion) kernels, with
+  //  each having its own build out to do.
+  //
+  // Current naming convention:
+  //
+  //  operands assumed in global memory : a, b
+  //
+  //  registers staging global load : ar, br (short for a/b read)
+  //
+  //  shared mem cache of operands : acw_smem, bcw_smem (short for a/b
+  //  cache_write smem)
+  //
+  //  registers at shared memory load output : acr, bcr (short for a/b cache
+  //  read)
+  //
+  //  register tensor input to the actual mma op: ab, bb (short for a/b
+  //  broadcasted)
+  //
+  //  accumulator register: cc (short for c cache)
+  //
+  //  result in global memory: c
+
+  // Currently only support a, b, c as fusion inputs/outputs
+  //  aka. no prolog and epilog fusion yet.
+  TORCH_CHECK(
+      c->isFusionOutput() && a->isFusionInput() && b->isFusionInput(),
+      "not supporting matmul fusion yet");
+  TORCH_CHECK(c->definition() && c->definition()->isA<MmaOp>());
+
+  mma_builder.configureMma(c);
+
+  // TODO:
+  // Beyond this point, mma_builder really just becomes a populated
+  //  list of parameters to describes the mma swizzles that should
+  //  be annotated on the tensor domain. Conceptually the mma builder
+  //  object should be separated to 2 parts, one as scheduler utility
+  //  and the other as matmul heuristic parameters, which we are
+  //  starting to build out.
+
+  // Setup register and shared memory stages:
+  //   TODO: this section goes to a separate matmul util,
+  //   and needs more configurability.
+
+  // Setup accumulator register.
+  auto cc = c->cacheBefore();
+
+  // Get the input to the mma op.
+  auto mma = dynamic_cast<MmaOp*>(cc->definition());
+  TORCH_INTERNAL_ASSERT(mma != nullptr);
+  auto ab = mma->inA()->as<TensorView>();
+  auto bb = mma->inB()->as<TensorView>();
+
+  // Get exact configurations from mma builder.
+  mma_builder.accumulatorTv(cc);
+  auto mma_options = mma_builder.build();
+
+  // Staging register for global memory load
+  TensorView *ar = a, *br = b;
+
+  if (!params.async_gmem_load_operands) {
+    ar = a->cacheAfter();
+    br = b->cacheAfter();
+  }
+
+  // TODO:
+  //  Significant build out needed here
+  //   for more flexibility and data type support.
+  // Shared memory
+  TensorView* acw_smem = nullptr;
+  TensorView* bcw_smem = nullptr;
+  // Shared memory read
+  TensorView* acr = nullptr;
+  TensorView* bcr = nullptr;
+
+  // Different paths because Volta swizzle needs to
+  //  involve the broadcast dimensions that are concretized
+  //  at mma, while Ampere ones should be done before
+  //  the broadcast op to be able to use cp.async.
```

Comment:
What's the limitation here? Is this because of our implementation or something more fundamental?

Reply:
Unfortunately, I have dived into documentation of both archs to tell what was the reason for this split. Shiming left only this comment here.

Reply:
Not a big deal, don't dive too deep right now ;-)